Source/SwiftLintCore/Extensions/Configuration+Parsing.swift

```diff
@@ -147,35 +152,104 @@ extension Configuration {
     }
 
     private static func validateConfiguredRulesAreEnabled(
+        parentConfiguration: Configuration?,
         configurationDictionary dict: [String: Any],
         ruleList: RuleList,
         rulesMode: RulesMode
     ) {
+        var enabledInParentRules: Set<String> = []
+        var disabledInParentRules: Set<String> = []
+        var allEnabledRules: Set<String> = []
+
+        // If our configuration is the default, precalculate some values
+        if case .default(let disabledRules, let optInRules) = rulesMode {
+            if case .only(let onlyRules) = parentConfiguration?.rulesMode {
+                enabledInParentRules = onlyRules
+            } else if case .default(let parentDisabledRules, let parentOptInRules) = parentConfiguration?.rulesMode {
+                enabledInParentRules = parentOptInRules
+                disabledInParentRules = parentDisabledRules
+            }
+            allEnabledRules = enabledInParentRules
+                .subtracting(disabledInParentRules)
+                .union(optInRules)
+                .subtracting(disabledRules)
+        }
+
         for key in dict.keys where !validGlobalKeys.contains(key) {
             guard let identifier = ruleList.identifier(for: key),
-                let rule = ruleList.list[identifier] else {
+                let ruleType = ruleList.list[identifier] else {
                     continue
             }
 
-            let message = "Found a configuration for '\(identifier)' rule"
-
             switch rulesMode {
             case .allEnabled:
                 return
-
             case .only(let onlyRules):
-                if Set(onlyRules).isDisjoint(with: rule.description.allIdentifiers) {
-                    Issue.genericWarning("\(message), but it is not present on '\(Key.onlyRules.rawValue)'.").print()
-                }
-
+                let issue = validateConfiguredRuleIsEnabled(onlyRules: onlyRules, ruleType: ruleType)
+                issue?.print()
             case let .default(disabled: disabledRules, optIn: optInRules):
-                if rule is any OptInRule.Type, Set(optInRules).isDisjoint(with: rule.description.allIdentifiers) {
-                    Issue.genericWarning("\(message), but it is not enabled on '\(Key.optInRules.rawValue)'.").print()
-                } else if Set(disabledRules).isSuperset(of: rule.description.allIdentifiers) {
-                    Issue.genericWarning("\(message), but it is disabled on '\(Key.disabledRules.rawValue)'.").print()
+                let issue = validateConfiguredRuleIsEnabled(
+                    parentConfiguration: parentConfiguration,
+                    enabledInParentRules: enabledInParentRules,
+                    disabledInParentRules: disabledInParentRules,
```

Comment:
The precalculated sets are only needed here, aren't they? If so, let's only compute them here when they are needed.